### PR TITLE
Target netcoreapp3.1

### PR DESCRIPTION
**Pull request by jaredpar**
This unifies our .NET Core assets to be targeting `netcoreapp3.1` where previously it was a mix of `netcoreapp2.1` and `netcoreapp3.0`.

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-infrastructure PTAL

File: src/Test/Utilities/Portable/FX/CappedStringWriter.cs [hunk 1]
```diff
@@ -31,6 +31,7 @@ public CappedStringWriter(int expectedLength)
                 _expectedLength = expectedLength;
                 _remaining = Math.Max(256, expectedLength * 4);
             }
+            _remaining = int.MaxValue;
```
Comment (cston) [inline src/Test/Utilities/Portable/FX/CappedStringWriter.cs]: >_remaining = int.MaxValue; [](start = 12, length = 26)

Is this additional intentional?

File: docs/infrastructure/cross-platform.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ The script will install .NET Core to `.dotnet` directory if it is not found on `
 
 ## Using the compiler
 
-After building, there will be a `csc` in the `artifacts/bin/csc/Debug/netcoreapp2.1` and `artifacts/bin/csc/Debug/net472` directories. Use the former to run on .NET Core and the latter on Mono.
+After building, there will be a `csc` in the `artifacts/bin/csc/Debug/netcoreapp3.1` and `artifacts/bin/csc/Debug/net472` directories. Use the former to run on .NET Core and the latter on Mono.
```
Comment (sharwell) [inline docs/infrastructure/cross-platform.md]: ❓ netcoreapp2.1 is LTS. Do we want to support both?

File: eng/build.ps1 [hunk 3]
```diff
@@ -608,10 +608,6 @@ try {
 
     $global:_DotNetInstallDir = Join-Path $RepoRoot ".dotnet"
     InstallDotNetSdk $global:_DotNetInstallDir $GlobalJson.tools.dotnet
```
Comment (tmat) [inline eng/build.ps1]: This should also be unnecessary.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 4]
```diff
@@ -5849,7 +5849,7 @@ public void BinaryFileErrorTest()
             CleanupAllGeneratedFiles(binaryPath);
         }
 
-#if !NETCOREAPP2_1
+#if !NETCOREAPP3_1
```
Comment (sharwell) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: 💡 You can use `NETCOREAPP` if you don't want to keep changing these in the future

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@
 using Xunit;
 
 // The ShadowCopyAnalyzerAssemblyLoader type is only defined for the below platforms
-#if NET472 || NETCOREAPP2_1
+#if NET472 || NETCOREAPP3_1
```
Comment (sharwell) [inline src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: 💭 Do we run this anywhere else?

File: src/Test/Utilities/Portable/FX/CappedStringWriter.cs (hunk 1 above)
Comment (jaredpar) [reply src/Test/Utilities/Portable/FX/CappedStringWriter.cs]: No. Thanks for catching. That was a temp change to help debug a failure.

File: src/Tools/BuildBoss/ProjectCheckerUtil.cs [hunk 6]
```diff
@@ -326,6 +326,7 @@ private bool CheckTargetFrameworks(TextWriter textWriter)
                     case "netcoreapp1.1":
                     case "netcoreapp2.1":
                     case "netcoreapp3.0":
```
Comment (sharwell) [inline src/Tools/BuildBoss/ProjectCheckerUtil.cs]: Do we still use these in the build? Consider removing any that we are no longer using.

File: docs/infrastructure/cross-platform.md (hunk 2 above)
Comment (jaredpar) [reply docs/infrastructure/cross-platform.md]: The compiler no longer ships to the LTS SDK by default: only to `netcoreapp3.1` and forward. Any changes we `netcoreapp2.1` would be a servicing event but it's incredibly unlikely the compiler will meet the servicing bar. Instead we'd fix in current SDK and tell customers to use new SDK to build for `netcoreapp2.1` (this is not special to Roslyn but instead just how our SDK layer works)

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Hmm I'd feel more comfortable with that if I can managee to remove our `netcoreapp1.1` dependency. Will try that in a follow up PR

File: src/Tools/BuildBoss/ProjectCheckerUtil.cs (hunk 6 above)
Comment (jaredpar) [reply src/Tools/BuildBoss/ProjectCheckerUtil.cs]: I think there are a few I can remove here.

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: Not sure what you mean by this. Can you elaborate a bit?

File: eng/build.ps1 (hunk 3 above)
Comment (jaredpar) [reply eng/build.ps1]: Removing this breaks most of our builds

```
Listing running build processes...
Downloading vswhere
Building bootstrap compiler
Downloading RoslynTools.MSBuild 16.4.0-alpha
F:\workspace\_work\1\s\.tools\msbuild\16.4.0-alpha\tools\MSBuild\Current\Bin\msbuild.exe /nologo /bl:F:\workspace\_work\1\s\artifacts\log\Debug\Bootstrap.binlog /consoleloggerparameters:Verbosity=minimal;summary /m /nodeReuse:false /p:TreatWarningsAsErrors=true /p:Configuration=Release /p:UseRoslynAnalyzers=false /p:ContinuousIntegrationBuild=true /p:RoslynEnforceCodeStyle=false /p:UseRoslynAnalyzers=false /p:DotNetUseShippingVersions=true /p:InitialDefineConstants=BOOTSTRAP /p:PackageOutputPath=F:\workspace\_work\1\s\artifacts\Bootstrap /p:EnableNgenOptimization=false /p:PublishWindowsPdb=false /restore /t:Pack /warnaserror src\NuGet\Microsoft.Net.Compilers.Toolset\Microsoft.Net.Compilers.Toolset.Package.csproj
  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download
F:\workspace\_work\1\s\src\NuGet\Microsoft.Net.Compilers.Toolset\Microsoft.Net.Compilers.Toolset.Package.csproj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.

Build FAILED.

F:\workspace\_work\1\s\src\NuGet\Microsoft.Net.Compilers.Toolset\Microsoft.Net.Compilers.Toolset.Package.csproj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
    0 Warning(s)
    1 Error(s)
```

Comment (tmat) [reply eng/build.ps1]: and replacing it with `InitializeDotNetCli` call?

File: eng/common/templates/job/performance.yml [hunk 7]
```diff
@@ -7,7 +7,7 @@ parameters:
   container: ''                   # required -- name of the container
   osGroup: ''                     # required -- operating system for the job
   extraSetupParameters: ''        # optional -- extra arguments to pass to the setup script
-  frameworks: ['netcoreapp3.0']   # optional -- list of frameworks to run against
+  frameworks: ['netcoreapp3.1']   # optional -- list of frameworks to run against
```
Comment (sharwell) [inline eng/common/templates/job/performance.yml]: ⚠️ This is arcade code /cc @tmat

Comment (jaredpar) [reply eng/common/templates/job/performance.yml]: Well ... I probably shouldn't change that then.

File: eng/build.ps1 (hunk 3 above)
Comment (jaredpar) [reply eng/build.ps1]: Tried that and it also failed. At this point want to separate this out from the PR.

File: src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: I'm asking if the `#if` necessary here. It seems to be enabled for all supported TFMs.

Comment (jaredpar) [reply src/Compilers/Core/CodeAnalysisTest/ShadowCopyAnalyzerAssemblyLoaderTests.cs]: Ah I see. Yeah this doesn't seem to make much sense now.

File: eng/build.ps1 (hunk 3 above)
Comment (tmat) [reply eng/build.ps1]: Sounds good.

**Comment by jaredpar** [comment]
@agocke you good here?

File: src/Workspaces/Core/MSBuild/Host/SimpleAnalyzerAssemblyLoaderService.cs [hunk 8]
```diff
@@ -12,7 +12,7 @@ internal sealed class SimpleAnalyzerAssemblyLoaderService : IAnalyzerService
     {
 #if NET472
         private readonly DesktopAnalyzerAssemblyLoader _loader = new DesktopAnalyzerAssemblyLoader();
-#elif NETCOREAPP1_1 || NETCOREAPP2_1
+#elif NETCOREAPP1_1 || NETCOREAPP3_1
```
Comment (agocke) [inline src/Workspaces/Core/MSBuild/Host/SimpleAnalyzerAssemblyLoaderService.cs]: wtf

**Comment by agocke** [comment]
Did we back away from the general "define a netcoreapp variable and use that in our #ifs"?

**Comment by sharwell** [comment]
@agocke The SDK defines `NETCOREAPP` if you don't need to be version specific

File: src/Workspaces/Core/MSBuild/Host/SimpleAnalyzerAssemblyLoaderService.cs (hunk 8 above)
Comment (jaredpar) [reply src/Workspaces/Core/MSBuild/Host/SimpleAnalyzerAssemblyLoaderService.cs]: Yeah I think I can delete the `NETCOREAPP1_1`

**Comment by jaredpar** [comment]
The general `NETCOREAPP` definition is the one I am going for. I want to follow up though that I can fully nuke `netcoreapp1.1` from our build tree.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 9]
```diff
@@ -343,7 +343,12 @@ public bool SkipCompilerExecution
 
         public string? TargetType
         {
-            set { _store[nameof(TargetType)] = CultureInfo.InvariantCulture.TextInfo.ToLower(value); }
+            set
+            {
+                _store[nameof(TargetType)] = value != null
+                    ? CultureInfo.InvariantCulture.TextInfo.ToLower(value)
+                    : null;
```
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs]: 💭 Is this not just `value?.ToLowerInvariant()`?

File: src/Compilers/Core/MSBuildTask/MvidReader.cs [hunk 10]
```diff
@@ -177,7 +177,7 @@ private static bool ReadUInt32(BinaryReader reader, out uint output)
             return true;
         }
 
-        private static bool ReadBytes(BinaryReader reader, int count, [NotNullWhen(true)] out byte[]? output)
+        private static bool ReadBytes(BinaryReader reader, int count, out byte[]? output)
```
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/MvidReader.cs]: extern alias to the one in Microsoft.CodeAnalysis?

**Pull request merged**